app/js/streaming/StreamController.js

```diff
@@ -298,12 +298,33 @@
             self.notify(MediaPlayer.dependencies.StreamController.eventList.ENAME_STREAMS_COMPOSED);
         },
 
+        onTimeSyncAttemptCompleted = function (e) {
+            var self = this,
+                manifest;
+
+            if (!e.error) {
+                // adjust the loadedTime of the manifest by the offset between
+                // the network time and the local time. this is because loadedTime
+                // is used in a number of places to work out availability etc.
+                manifest = self.manifestModel.getValue();
```

Comment:
> Any cases where that offset is not taken into account need to be fixed.

This hits the nail on the head. Most of your other concerns are where I might have hacked round existing issues rather than fix them. I'll rework my changes, and fix the problems I was avoiding.